### PR TITLE
THRIFT-4382: Replace the use of Indirect Object Syntax calls to new()

### PR DESCRIPTION
**Pull request by djzort**
Using Indirect Object Syntax to instantiate objects is not recommended as per https://perldoc.perl.org/perlobj.html#Invoking-Class-Methods

I have update the output to follow the recommended Class->new() syntax

### CODE REVIEW

**Comment by jeking3** [comment]
Could you rebase on master and refresh the PR so we get a clean build, now that CI is stable again?  Also we need a Jira ticket for the change:

https://thrift.apache.org/docs/HowToContribute

Thanks.

**Comment by jeking3** [comment]
Post the Jira issue (THRIFT-nnnn) in here when you have one as well - thanks.

**Comment by djzort** [comment]
I will get to it :)

**Comment by jeking3** [comment]
@djzort reminder we need a Jira THRIFT issue and squash and rebase on master.

**Comment by djzort** [comment]
I finally made a jira account, but i need a month to recover from the horrors

**Comment by djzort** [comment]
This should now comply with the contributors policy

**Comment by bforbis** [comment]
It would probably be a good idea to also update the code at `tutorial/perl/*` so that the main perl tutorial uses this preferred syntax.

**Comment by jeking3** [comment]
If you are able to finish this up and do the tutorial code, or if you are not able to complete this, please let me know.

**Comment by djzort** [comment]
it's still on my TODO list, but is very low. I will get to it sooner or later.

**Comment by djzort** [comment]
I've rebased, made further adjustments including to the examples, and repushed. CI is now running - lets see how it goes

**Comment by djzort** [comment]
Out of curiosity is there a 'maintainer' for the perl stuff? The coding style seems a little dated and has a non-perlish feel to the use of sigils (being pedantic)

**Comment by jeking3** [comment]
There isn't a dedicated perl maintainer for Thrift.  I have probably done most of the work recently in there.  Recent changes include better CPAN support, cross-test enhancements for multiplexed and graceful shutdown, namespaces and versioning related changes.

**Comment by djzort** [comment]
Hopefully it an be merged.

File: test/perl/TestServer.pl [hunk 1]
```diff
@@ -376,19 +382,21 @@ ()
...
   my $self = shift;
   my $arg = shift;
   print("testException($arg)\n");
-  if ($arg eq "Xception") {
-    die new ThriftTest::Xception({errorCode => 1001, message => $arg});
-  } elsif ($arg eq "TException") {
-    die "astring"; # all unhandled exceptions become TExceptions
-  } else {
-    return new ThriftTest::Xtruct({string_thing => $arg});
+  if ($arg eq 'Xception') {
```
[the remark this answers is not in the document]
Comment (djzort) [reply test/perl/TestServer.pl]: With eliminating useless interpolations of strings and swapping to postfix new's, I wanted to go the whole way and swap to vaguely Stroustrup style (K&R w/o Cuddling the else) which is arguably the most common style in perl

File: test/perl/TestServer.pl [hunk 2]
```diff
@@ -48,7 +48,7 @@
 $|++;
 
 sub usage {
-    print <<EOF;
+    print <<"EOF";
```
[the remark this answers is not in the document]
Comment (jeking3) [reply test/perl/TestServer.pl]: Okay, fair enough.

File: test/perl/TestServer.pl (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeking3) [reply test/perl/TestServer.pl]: Okay, fair enough.

**Comment by jeking3** [comment]
The feature test failed on perl tls/ssl which likely means some logic changed or there is a syntax error.  I am taking a look locally.

From https://api.travis-ci.org/v3/job/362963723/log.txt:
```
/usr/bin/python3 test/test.py --retry-count 5 --features .* --skip-known-failures --server cpp,c_glib,cl,d,java,csharp,py,py3,rb,hs,perl,php,go,nodejs,dart,erl,lua,rs,netcore

===============================================================================
*** Following 2 failures were unexpected ***:
If it is introduced by you, please fix it before submitting the code.
===============================================================================
server-client:          protocol:         transport:               result:
perl-tls                multi-binary      buffered-ip-ssl          failure(64)
perl-tls                binary            buffered-ip-ssl          failure(64)
===============================================================================
```

File: test/perl/TestServer.pl [hunk 3]
```diff
@@ -99,53 +99,59 @@ sub usage {
     exit 1;
 }
 
-my $handler = new ThriftTestHandler();
-my $handler2 = new SecondServiceHandler();
-my $processor = new ThriftTest::ThriftTestProcessor($handler);
-my $processor2 = new ThriftTest::SecondServiceProcessor($handler2);
+my $handler    = ThriftTestHandler->new();
+my $handler2   = SecondServiceHandler()->new;
```
Comment (jeking3) [inline test/perl/TestServer.pl]: This line is incorrect and causing the cross test feature failure.

File: test/perl/TestServer.pl [hunk 4]
```diff
@@ -99,53 +99,59 @@ sub usage {
...
+else {
+    $serversocket = Thrift::ServerSocket->new(\%opts);
 }
 my $transport;
 if ($opts{transport} eq 'buffered') {
-    $transport = new Thrift::BufferedTransportFactory();
-} elsif ($opts{transport} eq 'framed') {
-    $transport = new Thrift::FramedTransportFactory();
-} else {
+    $transport = Thrift::BufferedTransportFactory();
```
Comment (jeking3) [inline test/perl/TestServer.pl]: missing ->new()

File: test/perl/TestServer.pl [hunk 5]
```diff
@@ -99,53 +99,59 @@ sub usage {
...
+}
+elsif ($opts{transport} eq 'framed') {
+    $transport = Thrift::FramedTransportFactory();
```
Comment (jeking3) [inline test/perl/TestServer.pl]: missing ->new()

**Comment by jeking3** [comment]
After fixing those lines the feature test passes, so if you can fix that up we'll see how the build goes.

File: test/perl/TestServer.pl (hunk 3 above)
Comment (djzort) [reply test/perl/TestServer.pl]: I have just corrected this, and force pushed the updated commit

File: test/perl/TestServer.pl (hunk 4 above)
Comment (djzort) [reply test/perl/TestServer.pl]: I have just corrected this, and force pushed the updated commit

File: test/perl/TestServer.pl (hunk 5 above)
Comment (djzort) [reply test/perl/TestServer.pl]: I have just corrected this, and force pushed the updated commit

**Comment by jeking3** [comment]
```
root@31d4b8941d65:/thrift/src# git diff HEAD
diff --git a/test/perl/TestServer.pl b/test/perl/TestServer.pl
index 7f76060a..7257241e 100644
--- a/test/perl/TestServer.pl
+++ b/test/perl/TestServer.pl
@@ -99,8 +99,8 @@ if ($opts{ssl} and not defined $opts{cert}) {
     exit 1;
 }

-my $handler    = ThriftTest::ThriftTestHandler->new();
-my $handler2   = ThriftTest::SecondServiceHandler->new();
+my $handler    = ThriftTestHandler->new();
+my $handler2   = SecondServiceHandler->new();
 my $processor  = ThriftTest::ThriftTestProcessor->new($handler);
 my $processor2 = ThriftTest::SecondServiceProcessor->new($handler2);
```

**Comment by djzort** [comment]
Woops. Sorry for wasting your time. I have just pushed yet again.

**Comment by djzort** [comment]
In looking through which TestServer.pl isnt working right in tests, i couldnt help noticing someone didnt know how to take a list and join it in to a string with commas. So I simplified that mess.

**Comment by jeking3** [comment]
I rebased on master which passes CI (sometimes with a little help, but it passes).

**Comment by djzort** [comment]
> I rebased on master which passes CI (sometimes with a little help, but it passes).

Sorry I have been snowed under and unable to look at this more. But am looking at it now having been inspired by your comment.

**Comment by jeking3** [comment]
How bizarre.  I rebased on master like I did last time and it automatically closed it.  Please open another PR.

**Comment by djzort** [comment]
I am going to have to dissect it anyway, as for the life of me i cant see whats breaking the tests. Test test suite is decidedly non-perlish and i cant actually get it to run on a local machine - so i am having to rely on travis (etc) for feedback :(

**Comment by djzort** [comment]
Fixed and sent on https://github.com/apache/thrift/pull/1613

The issue was print and printf both have lists as arguments - but mixing them up goes poorly

**Pull request closed without merging**